### PR TITLE
Module redeclaration

### LINKED ISSUE #333: Multiple matching module declarations found for ${leftmostName}; please make module declarations unique
Hi  guys,
  
   Can the tsoa to support multiple same module definition ?

### PR DESCRIPTION
**Pull request by endor**
### All Submissions:

- [X] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [X] Have you written unit tests?
- [ ] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [X] This PR is associated with an existing issue?

closes #333

### CODE REVIEW

File: src/metadataGeneration/metadataGenerator.ts [hunk 1]
```diff
@@ -56,6 +56,34 @@ export class MetadataGenerator {
...
 
       ts.forEachChild(sf, node => {
+        /**
+         * If we declare a namespace within a module, like we do in `tsoaTestModule.d.ts`,
+         * we need to explicitly get the children of the module declaration
+         * (`declare module 'tsoaTest'`) - which are the moduleBlock statements,
+         * because otherwise our type resolver cannot iterate over namespaces defined in that module.
+         */
+        if (ts.isModuleDeclaration(node)) {
+          const moduleDeclaration = node as ts.ModuleDeclaration;
```
Comment (WoH) [inline src/metadataGeneration/metadataGenerator.ts:66]: TS should figure this out without explicit type casting

File: src/metadataGeneration/metadataGenerator.ts [hunk 2]
```diff
@@ -56,6 +56,34 @@ export class MetadataGenerator {
...
+          /**
+           * For some reason unknown to me, TS resolves both `declare module` and `namespace` to
+           * the same kind (`ModuleDeclaration`). In order to figure out whether it's one or the other,
+           * we check the name property. When it is a string literal, we assume that it's a
+           * module declaration, not a namespace. I haven't found any other differences between the
+           * objects, except `name` and `localSymbol`. Since `localSymbol` is not defined anywhere in
+           * the types, we fall back on the `name` checking.
+           */
+          if (moduleDeclaration.name.kind === ts.SyntaxKind.StringLiteral) {
```
Comment (WoH) [inline src/metadataGeneration/metadataGenerator.ts:75]: you can check based on the flags so we can avoid this

File: src/metadataGeneration/metadataGenerator.ts (hunk 1 above)
Comment (endor) [reply src/metadataGeneration/metadataGenerator.ts]: Fixed.

File: src/metadataGeneration/metadataGenerator.ts (hunk 2 above)
Comment (endor) [reply src/metadataGeneration/metadataGenerator.ts]: Fixed.

**Comment by WoH** [comment]
Code seems fine now, let's at least add some definition/component schema specs around and this should be good to go.

**Comment by endor** [comment]
> Code seems fine now, let's at least add some definition/component schema specs around and this should be good to go.

Sorry, I don't understand what this means.

File: tests/fixtures/controllers/getController.ts [hunk 3]
```diff
@@ -78,6 +80,11 @@ export class GetTestController extends Controller {
     return {} as GetterInterfaceHerited;
   }
 
+  @Get('ModuleRedeclarationAndNamespace')
+  public async getModuleRedeclarationAndNamespace(): Promise<TsoaTest.TestModel73> {
```
Comment (WoH) [inline tests/fixtures/controllers/getController.ts:84]: This method should make tsoa produce additional OpenAPI specification. We should test that this happens and the produced piece of OAS is sound.
https://github.com/lukeautry/tsoa/blob/master/tests/unit/swagger/schemaDetails3.spec.ts
https://github.com/lukeautry/tsoa/blob/master/tests/unit/swagger/definitionsGeneration/definitions.spec.ts

Comment (endor) [reply tests/fixtures/controllers/getController.ts:84]: Is this ok?

File: tests/unit/swagger/schemaDetails3.spec.ts [hunk 4]
```diff
@@ -1343,4 +1343,24 @@ describe('Definition generation for OpenAPI 3.0.0', () => {
       });
     });
   });
+
+  describe('module declarations with namespaces', () => {
+    it('should generate the proper schema for a model declared in a namespace in a module', () => {
+      const ref = specDefault.spec.paths['/GetTest/ModuleRedeclarationAndNamespace'].get?.responses['200'].content?.['application/json']['schema']['$ref'];
+      expect(ref).to.equal('#/components/schemas/TsoaTest.TestModel73');
+      expect(specDefault.spec.components.schemas?.['TsoaTest.TestModel73']).to.deep.include({
```
Comment (WoH) [inline tests/unit/swagger/schemaDetails3.spec.ts:1351]: There's a helper you can use here: `getComponentSchema('TsoaTest.TestModel73')`
Also, is there a reason you're not using equals here?

File: tests/unit/swagger/definitionsGeneration/definitions.spec.ts [hunk 5]
```diff
@@ -66,6 +66,7 @@ describe('Definition generation', () => {
 
   describe('Interface-based generation', () => {
     it('should generate a definition for referenced models', () => {
+      console.log(allSpecs[0].spec.definitions);
```
Comment (WoH) [inline tests/unit/swagger/definitionsGeneration/definitions.spec.ts:69]: I believe we should instead add a test similar to the OpenAPI 3 one (using `getDefinition('...')`) and remove the log.

File: tests/unit/swagger/schemaDetails3.spec.ts (hunk 4 above)
Comment (endor) [reply tests/unit/swagger/schemaDetails3.spec.ts]: Fixed. There wasn't really a reason. I just wanted to skip properties that were undefined, but it doesn't really make a difference.

File: tests/unit/swagger/definitionsGeneration/definitions.spec.ts (hunk 5 above)
Comment (endor) [reply tests/unit/swagger/definitionsGeneration/definitions.spec.ts]: Fixed and added.

Comment (endor) [reply tests/unit/swagger/definitionsGeneration/definitions.spec.ts]: Also without the `console.log` this time ;)

File: src/metadataGeneration/metadataGenerator.ts [hunk 6]
```diff
@@ -56,6 +56,31 @@ export class MetadataGenerator {
...
+           * the same kind (`ModuleDeclaration`). In order to figure out whether it's one or the other,
+           * we check the node flags. They tell us whether it is a namespace or not.
+           */
+          // tslint:disable-next-line:no-bitwise
+          if (ts.NodeFlags.Namespace !== (node.flags & ts.NodeFlags.Namespace)) {
+            const moduleBlock = node.body as ts.ModuleBlock;
+            if (moduleBlock === null || moduleBlock.kind !== ts.SyntaxKind.ModuleBlock) {
+              return;
+            }
+            moduleBlock.statements.forEach(statement => {
```
Comment (WoH) [inline src/metadataGeneration/metadataGenerator.ts:77]:
```suggestion
          if ((node.flags & ts.NodeFlags.Namespace) === 0 && node.body && ts.isModuleBlock(node.body)) {
            node.body.statements.forEach(statement => {
```

File: src/metadataGeneration/typeResolver.ts [hunk 7]
```diff
@@ -689,19 +689,21 @@ export class TypeResolver {
...
-      }
+      statements = Array.prototype.concat(
+        ...moduleDeclarations.map(declaration => {
+          if (ts.isEnumDeclaration(declaration)) {
+            return declaration.members;
+          } else {
+            const moduleBlock = declaration.body as ts.ModuleBlock;
+            if (moduleBlock === null || moduleBlock.kind !== ts.SyntaxKind.ModuleBlock) {
+              throw new GenerateMetadataError(`Module declaration found for ${leftmostName} has no body.`);
+            }
```
Comment (WoH) [inline src/metadataGeneration/typeResolver.ts:701]:
```suggestion
            if (!declaration.body || !ts.isModuleBlock(declaration.body)) {
              throw new GenerateMetadataError(`Module declaration found for ${leftmostName} has no body.`);
            }
            return declaration.body.statements;
```

File: src/metadataGeneration/metadataGenerator.ts (hunk 6 above)
Comment (endor) [reply src/metadataGeneration/metadataGenerator.ts]: Looks like I missed that before. Should be included now.

**Comment by WoH** [comment]
Added 2 suggestions, you can commit them as is and go ahead and merge afterwards. Missed these during my initial review, sorry.

**Comment by simllll** [comment]
I just ran into this too, is there any pre release so I can test if this fixes my issue?

**Pull request merged**